### PR TITLE
[ON HOLD] Don't take protocol fee for in-market limit orders

### PR DESCRIPTION
**Pull request by narcis96**
ON HOLD - will be split into several PRs:
- https://github.com/cowprotocol/services/pull/2246
- https://github.com/cowprotocol/services/pull/2250
- https://github.com/cowprotocol/services/pull/2278
- todo [move driver model]

# Description
We don't want to take protocol fees for limit orders with in-market price.

# Changes
We read the quote from the db and compare the order price with the quote price.

# Fixes
https://github.com/cowprotocol/services/issues/2092

### CODE REVIEW

File: crates/autopilot/src/database/quotes.rs [hunk 1]
```diff
@@ -16,6 +19,29 @@ impl Postgres {
...
+    pub async fn read_quotes(&self, auction: &Auction) -> Result<HashMap<OrderUid, Quote>> {
+        let mut quote_tasks = Vec::new();
+        for order in &auction.orders {
+            let mut ex = self.pool.acquire().await?;
+            let order_uid = ByteArray(order.metadata.uid.0);
+            let quote_task = tokio::spawn(async move {
+                database::orders::read_quote(&mut ex, &order_uid)
+                    .await
+                    .ok()?
+            });
```
Comment (MartinquaXD) [inline crates/autopilot/src/database/quotes.rs:32]: You usually only want to use `tokio::spawn()` if the runtime is supposed to start working on the future in the background without having to `.await` it.
In this case it makes more sense to build an interator that yields futures. Then you can await all the futures in parallel using `futures::future::join_all()`.
FYI it's possible to write the future iterator in a way that allows collecting into a hashmap.

File: crates/autopilot/src/database/quotes.rs [hunk 2]
```diff
@@ -16,6 +19,29 @@ impl Postgres {
...
+            quote_tasks.push((order.metadata.uid, quote_task));
+        }
+        let mut quotes_map = HashMap::new();
+        for (order_uid, quote_task) in quote_tasks {
+            let quote = quote_task
+                .await?
+                .ok_or(anyhow::anyhow!("failed to parse quote"))?;
```
Comment (MartinquaXD) [inline crates/autopilot/src/database/quotes.rs:39]: For turning an option or some other error into an anyhow error you can also use `.context("failed to parse quote")`

File: crates/autopilot/src/run_loop.rs [hunk 3]
```diff
@@ -493,14 +495,66 @@ impl RunLoop {
     }
 }
 
+/// Checks whether or not an order's limit price is outside the market price
+/// specified by the quote.
+///
+/// Note that this check only looks at the order's limit price and the market
+/// price and is independent of amounts or trade direction.
+pub fn is_order_outside_market_price(
```
Comment (MartinquaXD) [inline crates/autopilot/src/run_loop.rs:503]: Instead of duplicating this function you can take the original one and make it reusable for this use case as well.

File: crates/autopilot/src/shadow.rs [hunk 4]
```diff
@@ -201,6 +201,7 @@ impl RunLoop {
             self.score_cap,
             self.solve_deadline,
             self.fee_policy.clone(),
+            HashMap::new(),
```
Comment (MartinquaXD) [inline crates/autopilot/src/shadow.rs:204]: Would be cool if we could get the correct fee policies as part of the shadow competition as well but that is probably not straight forward since we don't have a DB connection in this case.

**Review by fleupold — COMMENTED** [review]
We are adding a lot of logic and dependencies to the autopilot runloop due to fees (we are using the run_loops db access to fetch quotes which we pass down into the solve request, we instantiate the run_loop with the fee policies that need to be applied, we have a specific function to compute the fee policies all inside the run_loop).

I wonder how good this will scale. What are the trade-offs of having some fee component which knows about its dependencies (ie a database to look up quotes, the fee policy arguments) and can `apply` the required fee policy logic to an order?

How do we build this in a way that scales for the other types of fees we want to take eventually? How do we prevent the run-loop from becoming our next "SettlementEncoder" or "SolutionSubmissionStrategy" (ie pieces of code no-one wants to touch anymore because they are so brittle )?

https://github.com/cowprotocol/services/pull/2157 has some attempt and more context on this discussion I believe.

**Comment by fleupold** [comment]
The other alternative I see could be to reuse `is_order_outside_market_price` where it is defined today and keep the market order order class label `Market` for "inside market price" limit orders. I'm not sure about the implications, maybe @sunce86 has some thoughts.

**Comment by sunce86** [comment]
> We are adding a lot of logic and dependencies to the autopilot runloop due to fees (we are using the run_loops db access to fetch quotes which we pass down into the solve request, we instantiate the run_loop with the fee policies that need to be applied, we have a specific function to compute the fee policies all inside the run_loop).
> 
> #2157 has some attempt and more context on this discussion I believe.

Yes, we should first merge #2157. 

> The other alternative I see could be to reuse is_order_outside_market_price where it is defined today and keep the market > order order class label Market for "inside market price" limit orders. I'm not sure about the implications, maybe @sunce86 > has some thoughts.

With regards to network fees, it is too early to do this because we would have `Market` order with `fee_amount = 0`. For now, I think it's fine to have `is_order_outside_market_price ` logic in the `fee::Policies` [link](https://github.com/cowprotocol/services/pull/2157/files#diff-a344797890ec4b593718bbe8bae9a1c4f637935430553091d625f4f4d784e765R19) and replace it later with order class label.

**Comment by fleupold** [comment]
> because we would have Market order with fee_amount = 0

What's the issue with this? I remember the frontend team asking us to provide the labels. Are we depending on the assumption `class::Market <=> fee_amount=0` elsewhere in the codebase?

**Comment by sunce86** [comment]
Wouldn't we end up charging 0 fee in that case?

File: crates/autopilot/src/run_loop.rs [hunk 5]
```diff
@@ -493,14 +495,66 @@ impl RunLoop {
...
+        .filter_map(|order| {
+            match order.metadata.class {
+                OrderClass::Market => None,
+                OrderClass::Liquidity => None,
+                // TODO: https://github.com/cowprotocol/services/issues/2115
+                // skip protocol fee for TWAP limit orders
+                OrderClass::Limit(_) => {
+                    let quote = quotes.get(&order.metadata.uid)?;
+                    let quote_buy_amount = big_decimal_to_u256(&quote.buy_amount)?;
+                    let quote_sell_amount = big_decimal_to_u256(&quote.sell_amount)?;
```
Comment (sunce86) [inline crates/autopilot/src/run_loop.rs:531]: I see `fee_policies` function as part of the domain logic of autopilot. It should not be clogged with conversion of types. Let's move this conversion outside by making sure that `read_quotes` returns some `autopilot::Quote` struct with already prepared types for working with, rather than a `database::Quote`.

**Review by sunce86 — COMMENTED** [review]
Logic lgtm. Let's just reorganize the code according to code review comments so it's more aligned with the ddd approach. Also, it needs to rebase after https://github.com/cowprotocol/services/pull/2157 is merged.

File: crates/autopilot/src/run_loop.rs (hunk 3 above)
Comment (narcis96) [reply crates/autopilot/src/run_loop.rs:503]: It is unfortunately not possible. 

The original function takes shared::order_quoting::Quote as argument.
shared::order_quoting::Quote is filled by the QuoteStoring which is read by crates/database/src/quotes.rs::get method. Not that the function is using quotes table.

On the other hand,  @sunce86 recommended me to use crates/database/src/orders.rs::read_quote (which returns database::orders::Quote).

Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:503]: You should be able to refactor `is_order_outside_market_price` to use only `U256` values and than reuse it in autopilot. I forsee the autopilot function will eventually be dropped and the check be done only in orderbook and populating the order class as @fleupold suggested (but can't be done now).

File: crates/autopilot/src/shadow.rs (hunk 4 above)
Comment (narcis96) [reply crates/autopilot/src/shadow.rs:204]: correct

Comment (sunce86) [reply crates/autopilot/src/shadow.rs:204]: @MartinquaXD do you maybe know why we don't have the access to db? for security reasons so we don't end up changing something?
I am worried that onboarding new solvers and checking the protocol fee mechanism will be imposible without db connection. OTOH, we do have staging if necesarry.

Comment (MartinquaXD) [reply crates/autopilot/src/shadow.rs:204]: I think when the `shadow` autopilot was written there was no need to store anything in the DB.
And over all I like that we don't need any DB connection for it to run.

File: crates/autopilot/src/run_loop.rs (hunk 3 above)
Comment (narcis96) [reply crates/autopilot/src/run_loop.rs:503]: that's correct. will do that !

File: crates/autopilot/src/shadow.rs (hunk 4 above)
Comment (narcis96) [reply crates/autopilot/src/shadow.rs:204]: so do you think is it worth to modify the code of the shadow to be able to read the order and compute the fee policies or maybe is not worth the investment ?

File: crates/autopilot/src/run_loop.rs (hunk 5 above)
Comment (narcis96) [reply crates/autopilot/src/run_loop.rs:531]: I see that https://github.com/cowprotocol/services/pull/2157 is merged now. So will do the rework and move to the new file.

File: crates/autopilot/src/shadow.rs (hunk 4 above)
Comment (MartinquaXD) [reply crates/autopilot/src/shadow.rs:204]: I think for testing fee policies we might have to add the original quote to the order (or some other indicator that fees should be applied) in `/auction`. However, **if** we decide to go that route we should do it in a separate PR.
With that taken care of the only reason the `shadow` autopilot would need a DB is to store the data we need before calling `/settle` which I think is outside of the scope of what the `shadow` environment was initially envisioned to do. Originally it was only supposed to be a test bed to onboard new solvers without risk.

File: crates/autopilot/src/run_loop.rs [hunk 6]
```diff
@@ -309,13 +309,15 @@ impl RunLoop {
 
     /// Runs the solver competition, making all configured drivers participate.
     async fn competition(&self, id: AuctionId, auction: &Auction) -> Vec<Participant<'_>> {
+        let quotes = self.database.read_quotes(auction).await.unwrap();
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/autopilot/src/run_loop.rs]: updated

File: crates/autopilot/src/database/quotes.rs [hunk 7]
```diff
@@ -16,6 +19,29 @@ impl Postgres {
         database::quotes::remove_expired_quotes(&mut ex, max_expiry).await?;
         Ok(())
     }
+
+    pub async fn read_quotes(&self, auction: &Auction) -> Result<HashMap<OrderUid, Quote>> {
+        let mut quote_tasks = Vec::new();
+        for order in &auction.orders {
+            let mut ex = self.pool.acquire().await?;
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/autopilot/src/database/quotes.rs]: fixed.

File: crates/autopilot/src/database/quotes.rs (hunk 1 above)
Comment (sunce86) [reply crates/autopilot/src/database/quotes.rs]: Will try with sequential call first and evaluate the performance

File: crates/autopilot/src/database/quotes.rs (hunk 2 above)
Comment (sunce86) [reply crates/autopilot/src/database/quotes.rs]: updated

File: crates/autopilot/src/run_loop.rs (hunk 3 above)
Comment (sunce86) [reply crates/autopilot/src/run_loop.rs]: refactored

File: crates/autopilot/src/database/quotes.rs [hunk 8]
```diff
@@ -16,6 +19,20 @@ impl Postgres {
...
     }
+
+    pub async fn read_quotes(&self, auction: &Auction) -> Result<HashMap<OrderUid, Quote>> {
+        let mut ex = self.pool.acquire().await?;
+        let mut quotes = HashMap::new();
+        for order in &auction.orders {
+            let order_uid = ByteArray(order.metadata.uid.0);
+            let quote = database::orders::read_quote(&mut ex, &order_uid)
+                .await?
+                .context("failed to read quote")?;
```
Comment (fleupold) [inline crates/autopilot/src/database/quotes.rs:30]: This can happen (cf. this [oncall issue](https://cowservices.slack.com/archives/C0371SB243E/p1703527547249929?thread_ts=1703462512.028329&cid=C0371SB243E)) so we should probably not fail the entire method for it

File: crates/autopilot/src/database/onchain_order_events.rs [hunk 9]
```diff
@@ -600,7 +600,12 @@ fn convert_onchain_order_placement(
         order_data.fee_amount
     };
     let is_outside_market_price = if let Ok(ref quote) = quote {
-        if is_order_outside_market_price(&order_data.sell_amount, &order_data.buy_amount, quote) {
+        if is_order_outside_market_price(
```
Comment (fleupold) [inline crates/autopilot/src/database/onchain_order_events.rs:603]: nit: something like `if !order.within_market(quote)` may be nicer to read

File: crates/autopilot/src/protocol/fee.rs [hunk 10]
```diff
@@ -7,44 +7,79 @@
...
-                        // skip protocol fee for limit orders with in-market price
-
                         // TODO: https://github.com/cowprotocol/services/issues/2115
                         // skip protocol fee for TWAP limit orders
                         OrderClass::Limit(_) => {
-                            Some((order.metadata.uid, vec![fee_policy_to_dto(&config)]))
+                            let quote = quotes.get(&order.metadata.uid)?;
+                            let quote_buy_amount = big_decimal_to_u256(&quote.buy_amount)?;
+                            let quote_sell_amount = big_decimal_to_u256(&quote.sell_amount)?;
+                            let is_in_money_order = !is_order_outside_market_price(
```
Comment (fleupold) [inline crates/autopilot/src/protocol/fee.rs:54]:
```suggestion
                            let is_market_order = !is_order_outside_market_price(
```

File: crates/autopilot/src/run_loop.rs [hunk 11]
```diff
@@ -310,7 +309,14 @@ impl RunLoop {
 
     /// Runs the solver competition, making all configured drivers participate.
     async fn competition(&self, id: AuctionId, auction: &Auction) -> Vec<Participant<'_>> {
-        let fee_policies = fee::Policies::new(auction, self.fee_policy.clone());
+        let fee_policies = self
+            .policy_factory
+            .build(auction)
+            .await
+            .unwrap_or_else(|err| {
```
Comment (fleupold) [inline crates/autopilot/src/run_loop.rs:316]: I think this error handling is a bit too strict. While I agree we should investigate whenever we cannot attach a fee to an order as planned, we don't need to nullify all fees for all other orders that worked out fine.

File: crates/autopilot/src/run_loop.rs [hunk 12]
```diff
@@ -552,7 +558,10 @@ pub fn solve_request(
                     class,
                     app_data: order.data.app_data,
                     signature: order.signature.clone(),
-                    fee_policies: fee_policies.get(&order.metadata.uid).unwrap_or_default(),
+                    fee_policies: fee_policies
+                        .get(&order.metadata.uid)
+                        .cloned()
+                        .unwrap_or_default(),
```
Comment (fleupold) [inline crates/autopilot/src/run_loop.rs:564]: Why aren't fee policies part of the `Auction` and auction specific order struct to begin with? 
Wouldn't it be reasonable to build the fee policies at the same time when we build the auction rather than putting it on the critical path of sending the solve request?

File: crates/autopilot/src/protocol/fee.rs [hunk 13]
```diff
@@ -7,44 +7,79 @@
...
+pub struct PolicyFactory {
+    config: arguments::FeePolicy,
+    db: Arc<Postgres>,
 }
 
-impl Policies {
-    pub fn new(auction: &Auction, config: arguments::FeePolicy) -> Self {
-        Self {
-            policies: auction
+impl PolicyFactory {
```
Comment (fleupold) [inline crates/autopilot/src/protocol/fee.rs:28]: nit: I'm not sure this is a true [factory pattern](https://en.wikipedia.org/wiki/Factory_method_pattern), given the create method is called `build` (is this maybe more of a [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern))?

I think this component should be made part of the auction building process and expose a single method that takes an auction without fees by value and returns an auction with fees applied.

**Review by fleupold — COMMENTED** [review]
Concern about the failability of the fee application and that it's part of the critical path of sending the auction rather than building the auction.

File: crates/autopilot/src/run_loop.rs (hunk 12 above)
Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:564]: Because then it would be a requirement to attach fee policies every time the order is created (because we use `model::Auction` directly instead of autopilot domain specific type) which means:
1. Fee policies have to be defined for both `orderbook` and `autopilot`
2. Fee policies are determined at the order creation time, so if `autopilot` and `orderbook` restart with new fee policies all existing orders have old policies and will be settled with old which seems weird, but maybe ok (so it depends on the expectation, it wouldn't be a bug but rather a feature).

IMO fee policies are strictly protocol related and should live in `autopilot` only. The proper way would be to move the fee policies building into autopilots`SolvableOrdersCache` where `SolvableOrdersCache` would have to return autopilot domain specific/enriched `Auction` type containing build fee policies but that requires significant refactor.

Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:564]: But if you think we should expand model::Order I can try that and see how far I come

File: crates/autopilot/src/protocol/fee.rs (hunk 13 above)
Comment (sunce86) [reply crates/autopilot/src/protocol/fee.rs:28]: I did not use any specific pattern, I just aimed to group all data (that does not change over time) needed to build the fee policies in one object.

File: crates/autopilot/src/database/quotes.rs (hunk 8 above)
Comment (sunce86) [reply crates/autopilot/src/database/quotes.rs]: You are right! Fixed error handling. Now the whole fee policy building fails only if the connection to db failed.

File: crates/autopilot/src/database/onchain_order_events.rs (hunk 9 above)
Comment (sunce86) [reply crates/autopilot/src/database/onchain_order_events.rs]: refactored.

File: crates/autopilot/src/protocol/fee.rs (hunk 10 above)
Comment (sunce86) [reply crates/autopilot/src/protocol/fee.rs]: renamed.

File: crates/autopilot/src/run_loop.rs (hunk 11 above)
Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:316]: Fixed in the policy factory.

File: crates/autopilot/src/protocol/fee.rs (hunk 13 above)
Comment (fleupold) [reply crates/autopilot/src/protocol/fee.rs:28]: I see, but by using nomenclature of such patterns future maintainers may think that a specific pattern was used/intended here. Can we call it FeeEngine or something less loaded in this case?

Also, this could be the first component of our autopilot domain (db should probably be an infra type that doesn't leak the fact that Postgres is used).

File: crates/autopilot/src/run_loop.rs (hunk 12 above)
Comment (fleupold) [reply crates/autopilot/src/run_loop.rs:564]: No I agree with your assessment that this should be part of the SolvableOrdersCache (and that should return the domain specific type).

If this is too much work to change right now we can continue converting the shared model into the domain specific model here. However, I'd still prefer letting the `FeePolicyBuilder` or `FeeEngine` operate on a core domain model (rather than defining new types that we glue together in the runloop).

Do you think we can have FeeEngine expose a method `fn apply_fees(auction: domain::Auction) -> domain::Auction` that we call before invoking `solveRequest(domain::Auction)` and which then simply converts the domain object into the dto?

So we would do something like
1.  solvable_orders -> shared::Auction
2. here: convert shared::Auction -> domain::Auction
3. apply_fee on domain::Auction
4. in solve_request domain::Auction -> dto

File: crates/model/src/order.rs [hunk 14]
```diff
@@ -350,6 +350,11 @@ impl OrderData {
             self.valid_to,
         )
     }
+
+    /// Checks if the order is a market order.
+    pub fn within_market(&self, quote_sell_amount: &U256, quote_buy_amount: &U256) -> bool {
```
Comment (fleupold) [inline crates/model/src/order.rs:355]: nit: you can extend OrderData inside the `order_quoting.rs` module to allow using the quote type as an argument here which makes this nicer to read.

```rs
pub trait OrderPricing {
    fn within_market(&self, quote: &Quote) -> bool;
}
impl OrderPricing for OrderData {
    /// Checks if the order is a market order.
    fn within_market(&self, quote: &Quote) -> bool {
        self.sell_amount.full_mul(quote.buy_amount) >= quote.sell_amount.full_mul(self.buy_amount)
    }
}
```

**Comment by sunce86** [comment]
Moved the policies into `domain`. Abstracted the `Database`. Still not happy with the `config` but don't want to go too deep and out of the scope of this PR.

File: crates/autopilot/src/domain/fee/policy.rs [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    },
+    std::{
+        collections::HashMap,
+        sync::{Arc, RwLock},
+    },
+};
+
+/// Protocol fee policies with cache being updated on each auction.
+#[derive(Debug)]
+pub struct PoliciesCache {
```
Comment (fleupold) [inline crates/autopilot/src/domain/fee/policy.rs:20]: I'm not a fan of adding this policies cache. A cache is usually used for time sensitive operations. In this case loading the quotes from the database may be the bottleneck, so I think if we wanted to build a cache it should be for the quote fetching operation (such a cache may be useful for other places). Then this component would use a cache instead of the database.

Letting the autopilot operate on a policy cache adds unnecessary conceptual complexity imo.

Fee policies just need to be applied or derived from a given order and I think this is exactly the interface the fee domain component should offer.

`fn(order) -> [FeePolicy]`

or event better

`fn(order) -> order` (return order with fee policies)

File: crates/autopilot/src/run_loop.rs [hunk 16]
```diff
@@ -555,7 +561,11 @@ pub fn solve_request(
                     class,
                     app_data: order.data.app_data,
                     signature: order.signature.clone(),
-                    fee_policies: fee_policies.get(&order.metadata.uid).unwrap_or_default(),
+                    fee_policies: fee_policies
+                        .get(&order.metadata.uid)
+                        .cloned()
+                        .map(|policies| policies.into_iter().map(FeePolicy::new).collect_vec())
+                        .unwrap_or_default(),
```
Comment (fleupold) [inline crates/autopilot/src/run_loop.rs:568]: nit: if you first unwrap or default you don't need the nested map I believe

File: crates/autopilot/src/domain/fee/policy.rs (hunk 15 above)
Comment (sunce86) [reply crates/autopilot/src/domain/fee/policy.rs]: Removed the caching feature. Тhe current interface is 
fn(orders) -> HashMap(order, policies) to reuse the same db connection for thousands of calls.

File: crates/autopilot/src/run_loop.rs (hunk 16 above)
Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:568]: nested map does conversion from domain to `driver` dto so I don't see how we can avoid it.

File: crates/autopilot/src/domain/fee/policy.rs (hunk 15 above)
Comment (fleupold) [reply crates/autopilot/src/domain/fee/policy.rs]: I see, the need for one db call to fetch all quotes makes sense. What happened to the idea of passing in an auction like domain model so that we can apply the fee policies inside this component? 

I'm still not very happy that we are attaching fee policies on the critical path (after the auction deadline has started) where all other "slow" prep work is happening in a background task. I understand it's not the easiest place to add logic, but I feel we are creating significant technical debt here.
Can we not create a different auction struct that is private to the autopilot domain and that allows adding to it before being returned from the SolvableOrdersCache?

We are loading all orders (without caching) from the db, so I think we can also do that with all quotes. This way we wouldn't need a complex `Arc<RWLock<>>` cache inside this component but can instead make its only focus be applying fee policies to auctions/orders. If certain DB queries (e.g quotes for order IDs) are cacheable I think we should have this logic in separate components.

File: crates/autopilot/src/infra/database/auction/postgres/mod.rs [hunk 17]
```diff
@@ -0,0 +1,16 @@
+use {crate::database::Postgres, anyhow::Result};
+
+pub mod dto;
+
+impl Postgres {
+    /// Saves the given auction to the database and returns the ID of the
+    /// auction.
+    pub async fn save(&self, auction: dto::Auction) -> Result<dto::AuctionId> {
```
Comment (squadgazzz) [inline crates/autopilot/src/infra/database/auction/postgres/mod.rs:8]: The function not only saves the Auction. Would it be cleaner to rename it?

File: crates/autopilot/src/solvable_orders.rs [hunk 18]
```diff
@@ -234,6 +228,8 @@ impl SolvableOrdersCache {
         let removed = counter.record(&auction.orders);
         filtered_order_events.extend(removed);
 
+        let auction =
```
Comment (fleupold) [inline crates/autopilot/src/solvable_orders.rs:231]: Why even create a `model::Auction` two lines above? I think we should just take the orders we have convert those into the domain model and then create the auction using the domain code (applying fees is domain logic).

Ideally we would get rid of `self.database.solvable_orders` above so this entire method can operate on domain models, but I assume this would make the refactor even larger. Can we capture this as a follow up task?

File: crates/autopilot/src/boundary/order.rs [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+            .into_iter()
+            .map(Into::into)
+            .collect(),
+        sell_token_balance: order.data.sell_token_balance,
+        buy_token_balance: order.data.buy_token_balance,
+        class: order.metadata.class.into(),
+        app_data: order.data.app_data,
+        signature: order.signature,
+        eth_flow: order.metadata.ethflow_data,
+        onchain_order: order.metadata.onchain_order_data,
```
Comment (fleupold) [inline crates/autopilot/src/boundary/order.rs:51]: Why is this needed?

File: crates/autopilot/src/domain/auction/order.rs [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    // on the first fill.
+    pub pre_interactions: Vec<Interaction>,
+    pub post_interactions: Vec<Interaction>,
+    pub sell_token_balance: boundary::SellTokenSource,
+    pub buy_token_balance: boundary::BuyTokenDestination,
+    pub class: Class,
+    pub app_data: boundary::AppDataHash,
+    pub signature: boundary::Signature,
+    pub eth_flow: Option<boundary::EthflowData>,
+    pub onchain_order: Option<boundary::OnchainOrderData>,
```
Comment (fleupold) [inline crates/autopilot/src/domain/auction/order.rs:31]: Shouldn't domain models define their own types if possible?

File: crates/autopilot/src/infra/database/auction/mod.rs [hunk 21]
```diff
@@ -0,0 +1,24 @@
+use {
+    self::postgres::dto,
+    super::Database,
+    crate::domain::{self},
+};
+
+pub mod postgres;
+
+impl Database {
```
Comment (fleupold) [inline crates/autopilot/src/infra/database/auction/mod.rs:9]: This should probably go into `infra::persistence`? It probably doesn't matter for the domain to know if the auction is stored in a database or using other storage medium

File: crates/autopilot/src/infra/solvers/dto/mod.rs [hunk 22]
```diff
@@ -0,0 +1,4 @@
+pub mod quote;
```
Comment (fleupold) [inline crates/autopilot/src/infra/solvers/dto/mod.rs:1]: Why are all these new files? Can't they be moved from existing dtos? Can we maybe leave that to a different PR (to keep the size of this one manageable)?

File: crates/autopilot/src/infra/database/auction/mod.rs [hunk 23]
```diff
@@ -0,0 +1,24 @@
+use {
+    self::postgres::dto,
+    super::Database,
+    crate::domain::{self},
+};
+
+pub mod postgres;
```
Comment (fleupold) [inline crates/autopilot/src/infra/database/auction/mod.rs:7]: I think the module structure should be `infra::persistence::postgres::dto::{auction, order, fee, quote}`

**Review by fleupold — COMMENTED** [review]
I think this is a great step in the right direction! Could you please create a separate PR for the refactoring that changes `SolvableOrdersCache.current_auction` to return the domain model (without any logic changes)?

And then add the fee policy and application logic to the domain model in a separate PR?

Moving `driver_model` into the new structure should also be a separate PR imo.

File: crates/autopilot/src/database/auction.rs [hunk 24]
```diff
@@ -90,23 +95,21 @@ impl Postgres {
...
+            let mut quotes = HashMap::new();
+            for order in &orders {
+                if let Some(quote) =
+                    database::orders::read_quote(&mut ex, &ByteArray(order.metadata.uid.0)).await?
+                {
+                    quotes.insert(order.metadata.uid, postgres::dto::into_domain(quote));
+                }
+            }
+            quotes
+        };
```
Comment (fleupold) [inline crates/autopilot/src/database/auction.rs:108]: If performance is of importance we can probably rewrite the solvable_orders query to also fetch the quotes in the same query.

Comment (sunce86) [reply crates/autopilot/src/database/auction.rs:108]: Would be nice to do this when solvable orders cache is refactored to work with `domain::Order` type directly.

Comment (sunce86) [reply crates/autopilot/src/database/auction.rs:108]: Implemented

**Comment by sunce86** [comment]
Closing as the last 4th part of this PR is tracked here: https://github.com/cowprotocol/services/issues/2286
